### PR TITLE
Refactor & Improve Marks

### PR DESCRIPTION
**Pull request by johannes-wolf**
**Changes:**
- Marks draw functions now return elements
- The mark tip/base offset is now calculated by using new "tip" and "base" anchors
- The marks length is auto determined

**Visual Changes:**
- Straight marks have no default extra padding anymore, can be set by via the mark-style

**New Features:**
- Users can provide custom marks using the new `register-mark()` API

```typst
#canvas({
  import draw: *

  register-mark("face", style => {
    circle((0,0), radius: .5, fill: yellow)
    arc((0,0), start: 180deg + 30deg, delta: 180deg - 60deg, anchor: "origin", radius: .3)
    circle((-.15, +.15), radius: .1, fill: white)
    circle((-.10, +.10), radius: .025, fill: black)
    circle((+.15, +.15), radius: .1, fill: white)
    circle((+.20, +.10), radius: .025, fill: black)

    anchor("tip", (+.5, 0))
    anchor("base", (-.5, 0))
  }, mnemonic: ":)")

  catmull((-3, 0), (-1,1), (1,-1), (3,0), mark: (
    end: "face", start: (symbol: ":)", flip: true, reverse: true), ))
})
```

grafik

Ping: @Jollywatt

### CODE REVIEW

File: src/mark.typ [hunk 1]
```diff
@@ -102,55 +103,114 @@
...
+
+  return (
+    tip: tip,
+    base: base,
+    center: center,
+    length: vector.dist(tip, base),
+    drawables: drawables,
+  )
+}
+
```
Comment (fenjalien) [inline src/mark.typ:213]: It took me a while to figure out exactly what this function does as it does several things at once. I'm not happy with the quality of this code but I don't know right now how to improve it...

File: src/mark-shapes.typ [hunk 2]
```diff
@@ -34,259 +34,277 @@
...
+  tip-style.length = style.length * (ratio / 100%)
+  if style.reverse {
+    anchor("tip", tip)
+  } else {
+    anchor("tip", vector.add(tip, vector.scale(dir, _calculate-tip-offset(tip-style))))
+  }
+  if style.reverse {
+    anchor("base", vector.sub(base, vector.scale(dir, _calculate-tip-offset(tip-style))))
+  } else {
+    anchor("base", base)
```
[the remark this answers is not in the document]
Comment (fenjalien) [reply src/mark-shapes.typ:75]: bit long but sure

File: src/mark-shapes.typ [hunk 3]
```diff
@@ -34,259 +34,277 @@
...
-  if not symbol in marks {
-    (symbol, reverse) = mnemonics.at(symbol)
+  symbol = ctx.marks.mnemonics.at(symbol, default: symbol)
+  if symbol in ctx.marks.marks {
+    return (ctx.marks.marks.at(symbol), (:))
   }
-  return (marks.at(symbol), reverse)
+
+  let (symbol, defaults) = mnemonics.at(symbol, default: (symbol, (:)))
+  assert(symbol in marks, message: "Unknown mark '" + symbol + "'")
```
[the remark this answers is not in the document]
Comment (johannes-wolf) [reply src/mark-shapes.typ:308]: No, if the `ctx` does not contain the mark, we search for a mnemonic and _then_ we fail if we have not found a mark.

File: src/draw/styling.typ [hunk 4]
```diff
@@ -30,3 +30,40 @@
 ///
 /// - stroke (stroke): Stroke style
 #let stroke(stroke) = set-style(stroke: stroke)
+
+/// Register a custom mark to the canvas
+///
+/// The mark should contain both anchors called **tip** and **base** that are used to determine the marks orientation. If unset both default to `(0, 0)`.
+/// An anchor named **center** is used as center of the mark, if present. Otherwise the mid between **tip** and **base** is used.
```
[the remark this answers is not in the document]
Comment (johannes-wolf) [reply src/draw/styling.typ:37]: The example shows how it works.

File: src/mark.typ (hunk 1 above)
Comment (johannes-wolf) [reply src/mark.typ]: This is not a helpful comment. I renamed the function to better reflect what its purpose is (rendering a mark and finding the marks special anchors).

**Review by fenjalien — APPROVED** [review]

**Pull request merged**